### PR TITLE
test: nox for local testing

### LINKED ISSUE #292: [Enh]: Add local testing framework
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

It's an internal nice to have

### Please describe the purpose of the new feature or describe the problem to solve.

Using tox or nox for local testing can lead to less surprises when opening PRs

### Suggest a solution if possible.

_No response_

### If you have tried alternatives, please describe them below.

Manual local testing

### Would you want to open a pull request?

yes

### Additional information that may help us understand your needs.

Is this of any interest? If so, do we have preference/opinion on which framework to use?

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [x] ✅ Test
- [x] 🐳 Other

## Related issues 

- Closes #292 

## Checklist

- [x] Code follows style guide (ruff)
- [ ] Tests added  
image

- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.

It is fairly minimal but I liked the tool, it is the first time I am using it though 🙈

### CODE REVIEW

File: tests/test_common.py [hunk 1]
```diff
@@ -46,10 +46,10 @@
 df_right_pandas = pd.DataFrame({"c": [6, 12, -1], "d": [0, -4, 2]})
 df_right_lazy = pl.LazyFrame({"c": [6, 12, -1], "d": [0, -4, 2]})
 
-if os.environ.get("CI", None):
+if os.environ.get("CI", None):  # pragma: no cover
     try:
         import modin.pandas as mpd
-    except ImportError:  # pragma: no cover
+    except ImportError:
```
Comment (FBruzzesi) [inline tests/test_common.py:52]: For whatever reason, coverage was failing here and on `test_str.py` for the same reason

File: noxfile.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        "tests",
+        "--cov=narwhals",
+        "--cov=tests",
+        f"--cov-fail-under={coverage_threshold}",
+        "--runslow",
+    )
+    session.run("pytest", "narwhals", "--doctest-modules")
+
+
+@nox.session(python=PYTHON_VERSIONS)  # type: ignore[misc]
```
Comment (FBruzzesi) [inline noxfile.py:24]: type ignore is due to the following error raised in pre-commit:

> noxfile.py:24: error: Untyped decorator makes function "pytest_coverage" untyped  [misc]

Same for all the decorators below

File: noxfile.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import nox
+from nox.sessions import Session
+
+nox.options.default_venv_backend = "uv"
+nox.options.reuse_venv = True
```
Comment (FBruzzesi) [inline noxfile.py:5]: Probably opinionated, it just gets significantly faster for me 😇

**Review by MarcoGorelli — APPROVED** [review]
sure, let's do this, thanks @FBruzzesi ! gives me a chance to learn nox

and i'll give you an A+ for the Emily in Paris meme, great choice

**Comment by FBruzzesi** [comment]
> Emily in Paris meme

I am confused!
RyanGoslingCriticsChoiceAwardsGIF

**Comment by MarcoGorelli** [comment]
sorry, wrong series 😳

**Pull request merged**